### PR TITLE
Add QR decomposition to cupy

### PR DESCRIPTION
**Pull request by rezoo**
Merge after #2271.
This PR aims to add the QR decomposition to cupy.

Note: this PR is based on #1402.

### CODE REVIEW

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 1]
```diff
@@ -1,9 +1,75 @@
...
+        A = numpy.dot(A, A.transpose())
+        self.check_L(A)
+        # np.linalg.cholesky only uses a lower triangle of an array
+        self.check_L(numpy.array([[1, 2], [1, 9]]))
+
+
+@unittest.skipUnless(
+    cusolver_enabled, 'Only cusolver in CUDA 8.0 is supported')
+@testing.gpu
+class TestQRDecomposition(unittest.TestCase):
```
Comment (unnonouno) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: How about using parameterized test?

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 2]
```diff
@@ -1,9 +1,75 @@
 import unittest
 
+import numpy
+
+from cupy.cuda import cusolver_enabled
```
Comment (unnonouno) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: Don't import variable directly. Instead import module.
```
from cupy import cuda

...

cuda.cusolver_enabled
```

**Comment by rezoo** [comment]
I fixed the points you have mentioned.

**Comment by unnonouno** [comment]
I merged #2271. Pleaser rebase master.

**Comment by rezoo** [comment]
I rebased the branch.

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 3]
```diff
@@ -28,3 +29,43 @@ def test_decomposition(self):
...
+        result_cpu = numpy.linalg.qr(a_cpu, mode=mode)
+        result_gpu = cupy.linalg.qr(a_gpu, mode=mode)
+        if isinstance(result_cpu, tuple):
+            for b_cpu, b_gpu in zip(result_cpu, result_gpu):
+                cupy.testing.assert_allclose(b_cpu, b_gpu, atol=1e-4)
+        else:
+            cupy.testing.assert_allclose(result_cpu, result_gpu, atol=1e-4)
+
+    def test_r_mode(self):
+        self.check_mode(numpy.random.randn(2, 3), mode='r')
```
Comment (unnonouno) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: How about use parameterized test?

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 4]
```diff
@@ -28,3 +29,43 @@ def test_decomposition(self):
...
+
+    @testing.for_float_dtypes(no_float16=True)
+    def check_mode(self, array, mode, dtype):
+        a_cpu = numpy.asarray(array, dtype=dtype)
+        a_gpu = cupy.asarray(array, dtype=dtype)
+        result_cpu = numpy.linalg.qr(a_cpu, mode=mode)
+        result_gpu = cupy.linalg.qr(a_gpu, mode=mode)
+        if isinstance(result_cpu, tuple):
+            for b_cpu, b_gpu in zip(result_cpu, result_gpu):
+                cupy.testing.assert_allclose(b_cpu, b_gpu, atol=1e-4)
```
Comment (unnonouno) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: We need to check if they have the same dtype.

**Comment by rezoo** [comment]
Fixed all the points.
In my environments the cusolver is correctly enabled, but #2487 would be useful where cuSolver is not enabled without openmp.

**Comment by unnonouno** [comment]
I confirmed that it passes the test. LGTM

**Pull request merged**